[AI Accelerators] Update torch.nn.functional multi_head_attention_forward(). Add pastpath

ynonaolga opened the pull request:
Summary: It presents as a fastpath where common execution paths are executed natively in Pytorch using torch._native_multi_head_attention. At present, the fast path is implemented in [torch.nn.MultiHeadAttention](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/activation.py).
Change replicates fastpath implementation in torch.nn.functional multi_head_attention_forward.
With some refactoring done to avoid duplication.

Test Plan: buck2 test @//mode/dev-nosan //caffe2/test:jit -- 'jit_multihead_attn_forward \(test_jit\.TestJit\)' test_functional_multi_head_attn_fast_path --print-passing-details

Differential Revision: D40821294

soumith commented:
@ynonaolga can you please edit the title and description to be readable.

ynonaolga commented on torch/nn/functional.py:
```diff
@@ -4990,6 +4990,227 @@ def multi_head_attention_forward(
           head of shape :math:`(num_heads, L, S)` when input is unbatched or :math:`(N, num_heads, L, S)`.
     """
+
+    if torch.overrides.has_torch_function((
```
Added all inputs

ynonaolga commented on torch/nn/functional.py:
```diff
@@ -4990,6 +4990,227 @@ def multi_head_attention_forward(
+                                             static_v,
+                                             average_attn_weights)
+
+def multi_head_attention_forward_impl(
```
Made function private

albanD replied:
What do you mean "without the forward path"?
Also, it is not because you use it as an implementation detail of a nn.Module that this has to be a new public API on torch.nn.functional.
You should either have `multi_head_attention_forward` or `multi_head_attention_forward_impl` public but not both. Also the `_impl` in the function name doesn't really make sense for a public API.

albanD commented on torch/nn/functional.py:
```diff
@@ -4990,6 +4990,229 @@ def multi_head_attention_forward(
           :math:`S` is the source sequence length. If ``average_attn_weights=False``, returns attention weights per
           head of shape :math:`(num_heads, L, S)` when input is unbatched or :math:`(N, num_heads, L, S)`.
     """
+
```
Is this whole code completely duplicated from the nn.Module implementation? That does not sound right. We should either need a single implementation of this or refactor it into a helper function.

albanD commented on torch/nn/functional.py:
```diff
@@ -4990,6 +4990,266 @@ def multi_head_attention_forward(
+                                         key_padding_mask,
+                                         attn_mask)
+    if not why_not_fast_path:
+        in_proj_weight_ = torch.empty((3 * embed_dim_to_check, embed_dim_to_check), device=query.device, dtype=query.dtype) \
```
What are these Tensors with un-initialized memory used for?

ynonaolga replied:
They are used in torch._native_multi_head_attention(), which permits Nullable in_proj_weight argument.
Additional tensors introduced to avoid interruption with later called _multi_head_attention_forward_impl.

albanD commented on torch/nn/functional.py:
```diff
@@ -4990,6 +4990,266 @@ def multi_head_attention_forward(
+                                             static_v,
+                                             average_attn_weights)
+
+def can_use_fastpath(
```
Prepend an underscore to this name to make it private.

albanD commented on torch/nn/functional.py:
```diff
@@ -4990,6 +4990,266 @@ def multi_head_attention_forward(
+    return why_not_fast_path
+
+
+def multi_head_attention_forward_impl(
```
Add an underscore to the name to make it private please: `_multi_head_attention_forward_impl`

albanD commented on torch/nn/functional.py:
```diff
@@ -4990,6 +4990,266 @@ def multi_head_attention_forward(
+    static_v: Optional[Tensor] = None,
+    average_attn_weights: bool = True,
+) -> Tuple[Tensor, Optional[Tensor]]:
+    r"""
```
Is this the same as the one above. If so, you can remove it and ask the user to look at the doc in the function above.
If there are only small differences, list only the differences here.

albanD reviewed (commented):
The diff is still very hard to read as there is no information explaining what this is trying to achieve.

ynonaolga committed:
[AI Accelerators] Update torch.nn.functional multi_head_attention_forward(). Add pastpath (#88241)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88241

The fastpath from torch.nn replicated in the torch.nn.functional multi_head_attention_forward().

Test Plan: buck2 test @//mode/dev-nosan //caffe2/test:jit -- 'jit_multihead_attn_forward \(test_jit\.TestJit\)' test_functional_multi_head_attn_fast_path --print-passing-details

Differential Revision: D40821294

fbshipit-source-id: a174a0f062c3743f1f5b088146afb9d51412eac1

ynonaolga committed:
Merge branch 'master' into export-D40821294

ynonaolga committed:
Merge branch 'master' into export-D40821294

ynonaolga commented:
Error syncing with internal repo, starting anew
